### ISSUE TITLE
Discrepancy between formula and code in VAE_kl_divergence_loss 

### ISSUE THREAD

**Issue by tr31zh** [AndrewSpano/Disentangled_Variational_Autoencoder#12]
Hello,

Problem found in file vae.py class VAE.
In line 348 the mathematical formula states:
`D_{KL}[q_{phi}(z` | x) || p_{theta}(x)] = (1/2) * sum(std + mean^2 - 1 - log(std))

In line 357 the transcription reads:
`kl_divergence` = (1 / 2) * torch.sum(torch.exp(std) + torch.square(mean) - 1 - std, axis=1)

Is the missing log at the end of the transcription a typo?

Best regards,

**Comment by AndrewSpano** [comment]
Hello @tr31zh

Thank you for reaching out. In the docstring of the `_kl_divergence_loss()` function, specifically at line 350, we note that the standard deviation (std) has been substituted with it's logarithm (log(std)) to improve numerical stability (aka to avoid very-small/NaN values for the std). This is also why in the same line, `torch.exp(std)` is used instead of `std`.

While optimizing `f(g(x))` is not the same as optimizing `f(log(g(x)))`, often times the output of the latter it very close to the former. 

It is also possible to apply the logarithm directly to the output of the std layer, to further improve numerical stability. We chose not to. If you want, for your implementation you can edit line 98 of the same file to

```python
log_std = torch.log(self.std_layer(encoded_input))

return mean, log_std
```

I would suggest you edit the other relevant parts of the code as well to maintain consistency.
